src/node/Launcher.ts

```diff
@@ -305,10 +305,20 @@ class FirefoxLauncher implements ProductLauncher {
 
     let temporaryUserDataDir = null;
 
-    if (
-      !firefoxArguments.includes('-profile') &&
-      !firefoxArguments.includes('--profile')
-    ) {
+    const profileArgIndex = firefoxArguments.findIndex((arg) => {
+      return ['-profile', '--profile'].includes(arg);
+    });
+
+    if (profileArgIndex !== -1) {
+      const profilePath = firefoxArguments[profileArgIndex + 1];
+      if (!profilePath) {
+        temporaryUserDataDir = await this._createProfile(extraPrefsFirefox);
+        firefoxArguments.push(temporaryUserDataDir);
+      } else {
+        const prefs = this.defaultPreferences(extraPrefsFirefox);
+        await this.writePreferences(prefs, profilePath);
```

Comment:
Any preferences in `user.js` will be copied by Firefox to `prefs.js` and just removing `user.js` will not be enough to get it reverted. That means that all the changes will remain post Puppeteer. 

Given that we tweak the profile for automation the user might indeed see differences in Firefox' behavior afterward. I see two options:

1. Create a backup of prefs.js and restore it during teardown, which will not always work like in cases the user hits `Ctrl+C`. Also changes to `prefs.js` will trigger different behaviors for various components in Firefox, and can cause other files in the profile to get modified too.

2. Create a copy of the user's profile in the temp folder and let Puppeteer remove it as usual. 

Only with option 2 the users profile will be kept identical and no traces from a usage with Puppeteer will remain in the original profile. But note that this might also have side-effects if multiple tests require the same profile and in worst cases steps as done by a previous test.

With geckodriver we actually make use of option 2 and never modify the original profile.

Reply:
Actually creating a temporary copy of the specified user profile causes [some of the user-data-dir tests to fail](https://github.com/puppeteer/puppeteer/blob/ad7f1de44f78df4a304c8e052935e53d4685530c/test/launcher.spec.ts#L216-L229) because the expected path isn't the same anymore.

@jschfflr as it looks like this might be a larger refactoring because the newly created temp directory will not be accessible by the consumer, and the Browser class doesn't have a property.

What do you suggest? Shall we go with a backup of pref.js for now and delete both .js files after Puppeteer finished? It would at least copy the behavior as what is used for Chrome by using the real user data dir.

Reply:
Ah damn, that's too bad. Let's go with your proposed solution then!

Reply:
I can do. What would be a good place to actually reset the prefs.js and delete user.js? Is https://github.com/puppeteer/puppeteer/blob/main/src/node/BrowserRunner.ts#L100 the right place so that it's always executed?

Reply:
Yes, that's probably the best place for it!